File: examples/vector_addition_with_error_check/openmp_offload/vector_addition.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <omp.h>

// Size of array
#define N 1048576


// Main program
int main()
{
    // Number of bytes to allocate for N doubles
    size_t bytes = N*sizeof(double);
    
    //declaring start_time and end_time
    double start_time, end_time;
    float elapsed_time;

    // Allocate memory for arrays A, B, and C on host
    double *A = (double*)malloc(bytes);
    double *B = (double*)malloc(bytes);
    double *C = (double*)malloc(bytes);

   

    // Fill host arrays A, B, and C
    for(int i=0; i<N; i++)
    {
        A[i] = 1.0;
        B[i] = 2.0;
        C[i] = 0.0;
    }

    start_time = omp_get_wtime();
    //Summing arrays A and B for C
    #pragma omp target map(to:A[:N], B[:N]) map(from:  C[:N])
    #pragma omp  teams 
    for(int i=0; i<N; i++)
    {
	    C[i] = A[i] + B[i];
    }
    end_time = omp_get_wtime();
    elapsed_time = end_time - start_time;

    // Verify results
    double tolerance = 1.0e-14;
    for(int i=0; i<N; i++)
    {
        if( fabs(C[i] - 3.0) > tolerance )
        { 
            printf("Error: value of C[%d] = %f instead of 3.0\n", i, C[i]);
            exit(-1);
        }
    }	

    // Free CPU memory
    free(A);
    free(B);
    free(C);

    printf("\n---------------------------\n");
    printf("__SUCCESS__\n");
    printf("---------------------------\n");
    printf("N                 = %d\n", N);
    printf("---------------------------\n\n");
    printf("Elapsed Compute time      = %1.5f\n", elapsed_time);

    return 0;
}

File: examples/vector_addition_with_error_check/openacc/vector_addition.f90
program vector_addition
        use omp_lib
        use, intrinsic   :: iso_c_binding, only : c_loc,  c_size_t, c_sizeof, c_int
        implicit none
       
       integer,parameter :: N = 1048576
       integer           :: i
       real(8), allocatable :: A(:)
       real(8), allocatable :: B(:)
       real(8), allocatable :: C(:)
       DOUBLE PRECISION  :: byte, elapsed_time
       real(8) :: tolerance, start_time, end_time 
    
       !Allocating memory for A, B and C  
     !  allocate(A(n))  !
       allocate(A(N*sizeof(byte)))   !!to be consist with c code, we could allocate this way
       allocate(B(N))  !*sizeof(byte)))       !!but it still has N  dimensions
       allocate(C(N))  !*sizeof(byte)))
       
       !Initiatizing A, B and C
       do i=1,N
         A(i) = 1.0
         B(i) = 2.0
         C(i) = 0.0
       enddo
   
       !Summing arrays A and B for C
       start_time = omp_get_wtime()
       !$acc data copyin(A,B)    !!copying arrays A and B to device
       !$acc parallel loop
       do i=1,N

         C(i) = A(i) + B(i)
       enddo
       !$acc end data
       
       end_time = omp_get_wtime()
       elapsed_time = end_time - start_time

       !Verify results
       tolerance = 1.0d-14
       do i = 1, N
        if (abs(C(i) - 3.0d0) > tolerance) then
            print *, "Error: value of C(", i, ") = ", C(i), " instead of 3.0"
            stop 1
        end if
       end do
       
       ! Freeing up memory
       deallocate(A,B,C) 

       print *, ""
       print *, "---------------------------"
       print *, "__SUCCESS__"
       print *, "---------------------------"
       print *, "N                 = ", N
       print *, "---------------------------"
       print *, "Elapsed Compute Time           = ", elapsed_time
       print *, ""
     
end program

